### DISCUSSION TITLE
Parentheses in citations with biblatex-apa

### DISCUSSION THREAD

**Discussion by sorsted** [Witiko/markdown#166] [Q&A] [answered]
According to the [`biblatex-apa` documentation](http://mirrors.ctan.org/macros/latex/contrib/biblatex-contrib/biblatex-apa/biblatex-apa.pdf) (section 4.1), APA dictates that citations inside parentheses should not contain extra parentheses. Therefore, it provides the command

> `\nptextcite[⟨prenote⟩][⟨postnote⟩]{⟨key⟩}⟨punctuation⟩`

which works like `\textcite`, but suppresses the parentheses.

Is there a way to get this behaviour to be supported with the `markdown` package, so that the following will render correctly? I am willing to use LuaTeX if that helps.

```latex
\documentclass{article}

\usepackage[american]{babel}
\usepackage{csquotes}
\usepackage[style=apa]{biblatex}

\usepackage[citations]{markdown}

\begin{document}

\begin{markdown}

Here is a reference with parentheses: @someauthor [p. 233].

(Here is a reference without parentheses: @someauthor [p. 233].)

\end{markdown}

\printbibliography

\end{document}
```

EDIT: I know that I can do `[Here is a reference without parentheses: @someauthor, p. 233]`, but this will not work for more complicated examples. For instance, I cannot do `(This follows from @someauthor and also from @someotherauthor.)` this way. The only thing I can currently do is `[This follows from @someauthor; and also from @someotherauthor.]`, which inserts an unwanted semicolon in the output.

EDIT: Note that it is probably best if parenthesis levels are counted for each paragraph rather than each line, so that the following works:

```latex
(This is another reference to
@someauthor,
p. 233)
```

It is probably also best if the parenthesis level counter is reset for every new paragraph so that a forgotten `)` doesn’t screw up the rest of the document.

EDIT: It is probably also best if this can be solved in a way that doesn’t break existing LaTeX syntax involving parentheses when this appears in hybrid mode.

**Comment by Witiko** [comment]
@sorsted This is related to [the text citation renderer][1]. The Markdown package provides [a default renderer for LaTeX][2], which tries to cover a couple use cases (BibTeX, Natbib, and BibLaTeX), but you need to adopt it to special cases such as `biblatex-apa`. Here, it should be sufficient to replace the `\textcites` command with `\nptextcites`:

``` tex
\usepackage{xpatch}
\xpatchcmd
  \markdownLaTeXBibLaTeXTextCitations
  {\textcites}%
  {\nptextcites}%
  {}%
  {}
```

Please note that default renderers are subject to change, so if you want to be sure that your code does not break in the long run (e.g. we might get rid of the `\markdownLaTeXBibLaTeXTextCitations` command in the future), you should copy the renderer definition from the Markdown package to your document/style file and do your changes there.

 [1]: https://witiko.github.io/markdown/#text-citations-renderer
 [2]: https://mirrors.ctan.org/macros/generic/markdown/markdown.pdf#page=227

**Reply by sorsted** [reply]
But what if I want to be able to use both `\parencite`, `\textcite`, and `\nptextcite` citations? And  this different from how the pandoc markdown citation syntax usually works, so a lot of manual adjustments will have to be made to an existing, format-agnostic markdown document. There is no way to do the replacement automatically, e.g. use Lua code to test whether or not the given citation occurs inside a pair of parentheses or not?

**Reply by Witiko** [reply]
Based on whether the parentheses were used, either [the parenthesized citation renderer][1] or [the text citation renderer][2] will be invoked. You can define each of them to do their own thing in (La)TeX.

 [1]: https://witiko.github.io/markdown/#parenthesized-citations-renderer
 [2]: https://witiko.github.io/markdown/#text-citations-renderer

**Reply by Witiko** [reply]
In the Markdown package, the syntax is separate from what TeX does with it. This gives you a lot of flexibility, but it also requires you to code in TeX to achieve the end result that you'd like. Thankfully, there are [default renderers][1] that you can copy and adjust.

 [1]: https://mirrors.ctan.org/macros/generic/markdown/markdown.pdf#page=227

**Reply by Witiko** [reply]
Ah, sorry, I mistook parentheses for brackets: No, there is no way to react to parentheses in your text.

**Reply by Witiko** [reply]
What you can do is write `[This follows from @someauthor; and also from @someotherauthor.]` and then redefine the parenthesized citation renderer to do what you'd like it to do. There is no way to differentiate `(Here is a reference without parentheses: @someauthor [p. 233].)` and `Here is a reference without parentheses: @someauthor [p. 233].` from within the Markdown package at the moment. There is a tentative idea to allow users to define syntax extensions in Lua, which could be used for that, but we are still in early stages of that effort. More on that in an upcoming TUGboat article.

**Reply by sorsted** [reply]
Okay, looking forward to that feature being added. Until then, I guess the only available solution is to hack into the way `markdown` handles the syntax `-@someauthor`, which would usually suppress the author name (though this seems to be disabled in the `markdown` package at the moment when using BibLaTeX). The best thing I could pull off was the following:

```latex
\usepackage{gobble}
\makeatletter
\def\@gobblethird#1#2#3{{#1}{#2}}
\def\markdownLaTeXBibLaTeXCitations#1#2#3#4#5{%
    \advance\markdownLaTeXCitationsCounter by 1\relax
    \ifnum\markdownLaTeXCitationsCounter>\markdownLaTeXCitationsTotal\relax
    \ifx#2-\relax
        \expandafter
        \expandafter
        \expandafter
        \nptextcites
        \@gobblethree
    \else
        \expandafter\parencites
    \fi
    #1[#3][#4]{#5}%
    \expandafter\gobblefive
    \fi\expandafter\markdownLaTeXBibLaTeXCitations\@gobblethird{#1[#3][#4]{#5}}{#2}%
}
\def\markdownLaTeXBibLaTeXTextCitations#1#2#3#4#5{%
    \advance\markdownLaTeXCitationsCounter by 1\relax
    \ifnum\markdownLaTeXCitationsCounter>\markdownLaTeXCitationsTotal\relax
    \ifx#2-\relax
        \expandafter
        \expandafter
        \expandafter
        \nptextcites
        \@gobblethree
    \else
        \expandafter\textcites
    \fi
    #1[#3][#4]{#5}%
    \expandafter\gobblefive
    \fi\expandafter\markdownLaTeXBibLaTeXTextCitations\@gobblethird{#1[#3][#4]{#5}}{#2}%
}
\makeatother
```

The slightly complicated implementation ensures that it is the presence of a minus in front of the *first* citation that solely determines whether to use parentheses or not, i.e. `[-@someauthor; @someotherauthor; @somerandomauthor]` will render with no parentheses.

Am I doing this correctly, or is this “bad style”/“bad practice”?

**Reply by Witiko** [reply]
That should work. However, if you want to harden your document against possible changes in default renderers, you should also copy the definition of `\markdownLaTeXCitationsCounter` and the `\markdownSetup{rendererPrototypes={cite={...}, textCite={...}}}`. This will ensure that you don't depend on the default renderers if they should change.

**Reply by Witiko** [reply]
The citation renderers haven't changed in a couple years, but it is good practice.

**Reply by sorsted** [reply]
Okay, I will do that.

But this got me thinking again: Suppose that I make this redefinition. Then automating this process would amount to the following: When `markdown` wants to interpret the markdown syntax of some piece of text, run some regex replacement on it that replaces `@` inside a balanced pair of parentheses by `-@`. And only *then* run the markdown interpreter. Is this not possible to do with the current version of the `markdown` package?

I realize that this will break syntax like `write me on [this email address](mailto:foo@bar.com)`, but I think I can live with that.

**Reply by Witiko** [reply]
You could certainly modify `markdown.lua` to do that for you in the `M.reader` object. Alternatively, you could also modify the buffering routine for `\begin{markdown} ... \end{markdown}` in `markdown.tex` and use e.g. the l3regex library from LaTeX3 to do what you are after, but this would be more difficult and would not extend to commands such as `\markdownInput`.

**Reply by Witiko** [reply]
Here is a first draft of what user syntax extensions might look like: https://github.com/Witiko/markdown/discussions/172. Feedback will be appreciated.

**Comment by sorsted** [comment, marked as answer]
Update: The good guys at TeX.SX have provided me with [the perfect solution](https://tex.stackexchange.com/a/654261/19809) which doesn’t even require us to break with the standard markdown citation syntax. Furthermore, this will work even if later we decide to instead use pandoc to convert the markdown document to LaTeX for further post-editing. @Witiko, please tell me if I did everything correctly and have the right amount of `\expandafter`'s. At least it compiles just fine.

```latex
% Compile with LuaTeX:
\begin{filecontents*}{find_parens.lua}
-- From Marcel Krüger's answer at https://tex.stackexchange.com/a/654261/80496

-- How to translate boolean results to TeX macros selecting the right arguments
local bool_to_token = {[false] = token.create'use_ii:nn', [true] = token.create'use_i:nn'}

-- The codepoints we have to look for
local left_paren = string.byte'('
local right_paren = string.byte')'

-- Some boilerplate to define an unexpandable command sequence from Lua
local id = luatexbase.new_luafunction'ifInParensTF'
token.set_lua('ifInParensTF', id, 'protected')
lua.get_functions_table()[id] = function()

  -- We want to look at the current list
  local head = tex.nest.top.head

  local level = 0
  -- and look at all unprocessed characters
  for _, char in node.traverse_char(head) do
    -- Track the nesting level
    if char == left_paren then
      level = level + 1
    elseif char == right_paren then
      if level > 0 then
        level = level - 1
      else
        -- Negative levels don't make sense, so we ignore the ) instead. Print a warning though to avoid surprises.
        texio.write_nl'WARNING: Unable to reliably determine if inside parentheses since the paragraphs contains an unmatched right parentheses.'
        texio.write_nl'The unmatched parentheses will be ignored.'
      end
    end
  end
  -- Finally insert a macro which will select the first or second argument dependingon whether level > 0
  token.put_next(bool_to_token[level > 0])
end
\end{filecontents*}
\begin{filecontents*}{ref.bib}

@online{texSX,
    author={{egreg et al.}},
    year={2022},
    title={{TeX Stack Exchange}},
    url={https://tex.stackexchange.com/},
}

\end{filecontents*}

\documentclass{article}

\usepackage[american]{babel}
\usepackage{csquotes}
\usepackage[style=apa]{biblatex}

\directlua{require'find_parens'}

\addbibresource{ref.bib}

\usepackage[citations]{markdown}

\NewDocumentCommand \autotextcites {}{%
  \ifInParensTF {%
    \nptextcites
  }{%
    \textcites
  }%
}

\NewDocumentCommand \autoparencites {}{%
  \ifInParensTF {%
    \nptextcites
  }{%
    \parencites
  }%
}

\makeatletter

\let\markdownLaTeXCitationsCounter=\relax
\newcount\markdownLaTeXCitationsCounter

\def\markdownLaTeXBibLaTeXAPACitations#1#2#3#4#5{%
  \advance\markdownLaTeXCitationsCounter by 1\relax
  \ifnum\markdownLaTeXCitationsCounter>\markdownLaTeXCitationsTotal\relax
    \autoparencites#1[#3][#4]{#5}%
    \expandafter\@gobbletwo
  \fi\markdownLaTeXBibLaTeXAPACitations{#1[#3][#4]{#5}}}
\def\markdownLaTeXBibLaTeXAPATextCitations#1#2#3#4#5{%
  \advance\markdownLaTeXCitationsCounter by 1\relax
  \ifnum\markdownLaTeXCitationsCounter>\markdownLaTeXCitationsTotal\relax
    \autotextcites#1[#3][#4]{#5}%
    \expandafter\@gobbletwo
  \fi\markdownLaTeXBibLaTeXAPATextCitations{#1[#3][#4]{#5}}}

\markdownSetup{rendererPrototypes = {
	cite = {%
		\markdownLaTeXCitationsCounter=1%
		\def\markdownLaTeXCitationsTotal{#1}%
		\expandafter\markdownLaTeXBibLaTeXAPACitations
		\expandafter{\expandafter}%
	},
	textCite = {%
		\markdownLaTeXCitationsCounter=1%
		\def\markdownLaTeXCitationsTotal{#1}%
		\expandafter\markdownLaTeXBibLaTeXAPATextCitations
		\expandafter{\expandafter}%
	}%
}}

\makeatother


\begin{document}

\begin{markdown}

This follows from @texSX.

(It also follows from @texSX.)

(It also follows from [@texSX].)

(It also follows from
@texSX
among other places.)

(It also follows from other resources. Notice that we forgot a closing parenthesis.

But @texSX is a good start.

\end{markdown}

\printbibliography

\end{document}
```

**Reply by Witiko** [reply]
Nice solution. Mind that your code still assumes that `\markdownLaTeXCitationsCounter` exists, which it may not in the future. You should have `\let\markdownLaTeXCitationsCounter=\relax` followed by `\newcount\markdownLaTeXCitationsCounter` in your code.

**Reply by sorsted** [reply]
Like this?

**Reply by sorsted** [reply]
I noticed that the manual uses `renderers =` rather than `rendererPrototypes =`. Which is more correct here?
